File: go/parsers/yield.go
package parsers

import (
	"capnproto.org/go/capnp/v3"
	"github.com/xconnio/wampproto-go/serializers"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Yield struct {
	gen *gen.Yield
	ex  *PayloadExpander
}

func NewYieldFields(g *gen.Yield, payload []byte) messages.YieldFields {
	return &Yield{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.PayloadSerializerID()},
	}
}

func (y *Yield) RequestID() uint64 {
	return y.gen.RequestID()
}

func (y *Yield) Options() map[string]any {
	return map[string]any{}
}

func (y *Yield) Args() []any {
	return y.ex.Args()
}

func (y *Yield) KwArgs() map[string]any {
	return y.ex.Kwargs()
}

func (y *Yield) PayloadIsBinary() bool {
	return true
}

func (y *Yield) Payload() []byte {
	return y.ex.Payload()
}

func (y *Yield) PayloadSerializer() uint64 {
	return y.gen.PayloadSerializerID()
}

func YieldToCapnproto(m *messages.Yield) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	yield, err := gen.NewRootYield(seg)
	if err != nil {
		return nil, err
	}

	yield.SetRequestID(m.RequestID())
	payloadSerializer := selectPayloadSerializer(m.Options())
	yield.SetPayloadSerializerID(payloadSerializer)

	var payload []byte
	if m.PayloadIsBinary() {
		payload = m.Payload()
	} else {
		payload, err = serializers.SerializePayload(payloadSerializer, m.Args(), m.KwArgs())
		if err != nil {
			return nil, err
		}
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeYield, data, payload), nil
}

func CapnprotoToYield(data, payload []byte) (*messages.Yield, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	yield, err := gen.ReadRootYield(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewYieldWithFields(NewYieldFields(&yield, payload)), nil
}

File: python/wamp_msgs_capnp/parsers/yield_.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages import yield_ as yield_message
from wampproto.serializers.payload import serialize_payload

from wamp_msgs_capnp.parsers import helpers

# Load the Cap'n Proto schema
root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "yield.capnp")
yield_capnp = capnp.load(str(schema_file))


class Yield(yield_message.IYieldFields):
    def __init__(self, gen, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.payloadSerializerID)

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def options(self) -> dict[str, Any]:
        return {}

    @property
    def args(self) -> list[Any]:
        return self._ex.args

    @property
    def kwargs(self) -> dict[str, Any]:
        return self._ex.kwargs

    @property
    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes:
        return self._ex.payload

    @property
    def payload_serializer(self) -> int:
        return self._gen.payloadSerializerID


def yield_to_capnproto(y: yield_message.Yield) -> bytes:
    yield_obj = yield_capnp.Yield.new_message()

    yield_obj.requestID = y.request_id

    payload_serializer = helpers.select_payload_serializer(y.options)
    yield_obj.payloadSerializerID = payload_serializer

    if y.payload_is_binary():
        payload = y.payload
    else:
        payload = serialize_payload(payload_serializer, y.args, y.kwargs)

    packed_data = yield_obj.to_bytes_packed()

    return helpers.prepend_header(yield_message.Yield.TYPE, packed_data, payload)


def capnproto_to_yield(data: bytes) -> yield_message.Yield:
    message_data, payload_data = helpers.extract_message(data)
    yield_obj = yield_capnp.Yield.from_bytes_packed(message_data)

    return Yield(yield_obj, payload_data)
